File: 27 08 25/removenode.rb
# Definition for singly-linked list
class ListNode
  attr_accessor :val, :next
  def initialize(val = 0, _next = nil)
    @val = val
    @next = _next
  end
end

class Solution
  def remove_nth_from_end(head, n)
    dummy = ListNode.new(0, head)
    slow = dummy
    fast = dummy

    # Move fast ahead by n+1 steps
    (n + 1).times do
      fast = fast.next
    end

    # Move both pointers until fast reaches the end
    while fast
      slow = slow.next
      fast = fast.next
    end

    # Remove nth node
    slow.next = slow.next.next

    dummy.next
  end
end

# Helper function to build linked list from array
def build_list(arr)
  dummy = ListNode.new
  curr = dummy
  arr.each do |val|
    curr.next = ListNode.new(val)
    curr = curr.next
  end
  dummy.next
end

# Helper function to print linked list
def print_list(head)
  while head
    print "#{head.val} "
    head = head.next
  end
  puts
end

# Example usage
head = build_list([1, 2, 3, 4, 5])
solution = Solution.new
new_head = solution.remove_nth_from_end(head, 2)
print_list(new_head) # Output: 1 2 3 5

File: 2025/August/27/removenode.scala
// Definition for singly-linked list
class ListNode(var x: Int = 0, var next: ListNode = null)

object Solution {
  def removeNthFromEnd(head: ListNode, n: Int): ListNode = {
    val dummy = new ListNode(0, head)
    var slow: ListNode = dummy
    var fast: ListNode = dummy

    // Move fast ahead by n+1 steps
    for (_ <- 0 to n) {
      fast = fast.next
    }

    // Move slow and fast together
    while (fast != null) {
      slow = slow.next
      fast = fast.next
    }

    // Remove nth node
    slow.next = slow.next.next

    dummy.next
  }

  // Helper: build list from Scala List
  def buildList(arr: List[Int]): ListNode = {
    val dummy = new ListNode()
    var curr = dummy
    for (v <- arr) {
      curr.next = new ListNode(v)
      curr = curr.next
    }
    dummy.next
  }

  // Helper: print list
  def printList(head: ListNode): Unit = {
    var curr = head
    while (curr != null) {
      print(curr.x + " ")
      curr = curr.next
    }
    println()
  }

  def main(args: Array[String]): Unit = {
    val head = buildList(List(1, 2, 3, 4, 5))
    val newHead = removeNthFromEnd(head, 2)
    printList(newHead) // Output: 1 2 3 5
  }
}
